### DISCUSSION TITLE
C# preprocessor directives  #pragma exception "disable" "MyException" <<< new

### DISCUSSION THREAD

**Discussion by srburton** [dotnet/csharplang#1936] [General]
It would be interesting if you had a way to ignore exception without needing the try {} cath (Exception) {} By putting this policy.

```

 public Main(){
      
    #pragma exception disable MyException
     var value = Validate("string is valid?");
     #pragma exception disable MyException
    
     if(value is string)
      {
         ///....
      }
 }

 public string Validate(string value)
 {    
         throw new MyException("String invalid!");
 }

```

**Comment by YairHalberstadt** [comment]
Presumably the second #pragma was meant to be enable?

I understand it would be interesting, but what advantage would it provide?

**Comment by franchyd** [comment]
This would essentially be a syntactic sugar for 
```csharp
   try
   { 
      Validate();
   }
   catch (MyException) 
   {
   } 
```

 As exceptions are runtime beasts and preprocessor directives are not.  I'm not sure why you would need this. Or why you would do it with preprocessor directives.

**Comment by HaloFour** [comment]
This would codify and encourage very poor exception handling practices.

Even if this functionality were added your example still couldn't compile as `value` could never be considered definitely assigned.

**Comment by srburton** [comment]
@HaloFour  
This was a simple example to be easy to understand follows a practical example


**Situation number 1**
```
 public Main(){
      

     #pragma exception disable MyException
     var value = Validate("c# is good!");
     #pragma exception disable MyException
    
     if(value is string)
      {
         Console.Write(value);
      }
    
    //My code keeps running ....
     

 }

 public string Validate(string value)
 {    
        if(value != "c# is good!")
       {
               throw new MyException("String invalid!");
        }
     return  "Hello Word";
 }
//Output 
Hello Word
```



**Situation number 2**
```
 public Main(){
      
     try{
  
     var value = Validate("c# is good!"); 
  
     if(value is string)
      {
         Console.Write(value);
      }
      else
      {
           Console.Write("Invalid!");
       }

      //My code does not continual:-(

     }cath(Exception e)
     {
        Console.Write(e.Message);
     }
     //Continue code ...
   
     try{
           ///Another treatment block 
       }
        cath(Exception)
       {

       }

     
 }

 public string Validate(string value)
 {    
        if(value == "c# is good!")
       {
               throw new MyException("String invalid!");
        }
     return  "Hello Word";
 }

//Output 
String invalid!
```

**Conclusion**

**Comment by HaloFour** [comment]
@srburton 

The C# compiler can't stop exceptions from happening.  At most it could silently handle the exceptions at the call-site.  This behavior that you're requesting would be impossible to accomplish without CLR changes, and considering that it effectively breaks every rule that exists regarding good exception handling that's certainly never going to happen.

**Comment by Austin-bryan** [comment]
@srburton I'm curious as to why you think C# should add a feature that will allow for horrible horrible coding practices and lead to even more bugs. 

```csharp
#pragma disable exception
int x = GetX();
#pragma enable exception
```

The exception would've been thrown, but the `#pragma` disabled it. So what is the value of `x` now? is it 0, `null` (even though it's not nullable)`, what else could it be? Or, is `x` even defined anymore? This will only cause more problems in an excuse for lazy programming.

**Comment by srburton** [comment]
@HaloFour  
I am pleased he handle the exception silently, we have to think so the language has to be the broadest in itself and let the programmer decide whether to use or not just a sugar as the keyword, if this function already deal silently without needing a try /cath and the code keep spinning would be great,

**Comment by srburton** [comment]
@willard720   The value of the variable would be `null`

**Comment by yaakov-h** [comment]
So any not-definitely-assigned variable has a value of `default(T)`? This sounds like a great vector for bugs.

**Comment by bondsbw** [comment]
I believe #1151 would provide the framework for a much better approach:

```c#
Suppress<MyException>
{
    var value = Validate("string is valid?");
    if(value is string)
    {
        Console.Write(value);
    }
}
```
elsewhere...
```c#
public static void Suppress<TException>(Action a) where TException : Exception
{
    try { a(); }
    catch (TException) {}
}
```

**Comment by markusschaber** [comment]
I don't see any reason to invent syntactic sugar via pragmas for what is already available in the language by using `try ... catch`. It's a rare use case to really do nothing (in most cases, you will in fact want to at least log the exception, at least on trace / verbose levels), and they're better written out explicitly.

Even, the two `#pragma`s are more to type than the existing `try ... catch` construct... 👎

**Comment by stuartstein777** [comment]
> 
> 
> @willard720 The value of the variable would be `null`

It's an int...

**Comment by masonwheeler** [comment]
Is this dependent on another proposal that I haven't seen yet, to add a `try/cath` construct to the language? :stuck_out_tongue:

**Comment by jveselka** [comment]
You can write yourself a helper method
```C#
public TOut TryOrDefault<TException, TOut>(Func<TOut> func) where TException : Exception
{
    try {
        return func();
    } catch (TException) {
        return default;
    }
}
```
Usage does not look that bad compared to pragmas. It could be even improved a bit by #1349  
```C#
var value = TryOrDefault<MyException, string>(() => Validate("string is valid?"));
```

**Comment by srburton** [comment]
My idea of #pragma it would be because I like c # developer should have full control the language can't stop me, disable functions that in my business suit not alias rule c # should have a form in run mode can create primitive types and tam bém preprocessor directives so linguem be fully templatable, leave the default the uptight to framework, I love developing hack, virus, spam, and with the c # runtime making it possible to mold that would be an enormous advance as the extension that was a great fucker, now imagine the power of programming c # behavior with the own c # code time. Thinking not in an MVC project or Windows Form more in specific projects and can create .dll that contains these mutants of behavior, this may came to security problem and not rely too much on nuget can get around with signatures that have .dll code that might interfere with the c # mode thinking that it would be interpresantes to disable some features and Frameworks visual ja talk that the framework of support, would be very similar to the operator more with everything from c #.

**Comment by HaloFour** [comment]
@srburton 

You're already free to do the wrong thing in your programs by wrapping each statement in a `try`/`catch` and ignoring any exceptions thrown.  It's made difficult on purpose; it's bad practice and should be actively discouraged.

C# as a language can't modify how the framework or any compiled assembly works.  If it throws an exception, an exception will be thrown.  The most you, as a C# developer, get to decide is how to handle that exception when it reaches your code.  This really is any different from any other language.  And this is a good thing as if by some weird compiler option you could completely alter the behavior of the entire ecosystem that would make everything incredibly fragile.

And if your reason for wanting a feature like this is to make it easier to write sloppy malware then you're very unlikely to find any support for any proposal.

**Comment by srburton** [comment]
@HaloFour 

It's not language nature decide what I'll do with it or not, I might as well create malware or even make kernel for IOT with c # and can handle your core ai will programmer what it will do is not of character of language and of the laws of your parents , Only exists because you forgot to treat the language makes malware she also makes the vaccine soon says language is a tool who decides what's going to do with it is the programmer, it would be very useful I can create my #pragma or something else c #

**Comment by HaloFour** [comment]
@srburton 

> It's not language nature decide what I'll do with it or not,

Yes it is, by not including bad features.

I'm not on the language team nor do I speak for them, but I can guarantee you that this proposal will never happen.

**Comment by franchyd** [comment]
I'm also not sure that a well defined implementation of this is even possible.  It would need to function as a sort of code generation automatically inserting try-catches but as @HaloFour mentioned any variables declared in these sections would throw wrenches into generating valid code.  And even if it could generate valid code, it would be exceedingly difficult for the code following these sections to operate without error considering the unknown state of the variables.

**Comment by YairHalberstadt** [comment]
To my mind a `#pragma disable exception` would do more than just add a try catch clause. 

It would require a runtime change that would ignore any exception being thrown, and just continue execution straight on.

So call Math.Sqrt on a negative number? It'll loop forever.

Pass an index to an array that's greater than the length of the array? You access some random uninitialised data.

In other words, this is one of the stupidest, most dangerous ideas I have ever seen. And completely infeasible as well.

**Reply by cinnamonwolfy** [reply]
in other words, he wants to drop the sharp from csharp

...

i will show myself out

**Comment by masonwheeler** [comment]
@YairHalberstadt `On Error Resume Next` ?

**Comment by srburton** [comment]
As we can see in this case it would be nice to give a ignore exception at IDE level because it is a controllable error.

![screenshot_2](https://user-images.githubusercontent.com/15394508/47241967-d5920480-d3c3-11e8-8b10-d882818c61a8.jpg)

**Comment by franchyd** [comment]
Are you suggesting you just want to disable debugger breakpoints for exceptions within your segments?

**Comment by srburton** [comment]
Yes, I want to because I already have treatment on my cath right down then that ignore would throw the exception for the previous treatment...

**Comment by CyrusNajmabadi** [comment]
@srburton i can't parse that.  Can you repeat what you're trying to say?

**Comment by CyrusNajmabadi** [comment]
You're in a delegate.  The surrounding try/catch has no bearing on what goes on in that delegate.

**Comment by srburton** [comment]
Now that I'm in time to explain my idea would be basically this that from the beginning I want to communicate.
![screenshot_3](https://user-images.githubusercontent.com/15394508/47242692-ef344b80-d3c5-11e8-8ef2-8ba3118125a0.jpg)

**Comment by CyrusNajmabadi** [comment]
@srburton i have no idea how that would work.  Why would the first exception go to the first catch block, and the second exception would go to the second catch block?

**Comment by srburton** [comment]
This is the normal situation that we all know today, I did a treatment with an alert ja and when I give an exception within a delegate that execption of a warning in Visual Studio the #pragma would be to disable this warning that Visual Studio and continue the  Code because it is an exception that has already been handled 

![normal_situation](https://user-images.githubusercontent.com/15394508/47243054-3a029300-d3c7-11e8-9264-13c157a09ac0.jpg)


Sugestion: 



![action_situation](https://user-images.githubusercontent.com/15394508/47243160-98c80c80-d3c7-11e8-9dee-b820527722b4.jpg)

**Comment by srburton** [comment]
When placed #pragma he simply ignores continues because he knows he has a treatment before breaking the code.

**Comment by CyrusNajmabadi** [comment]
> When placed #pragma he simply ignores continues because he knows he has a treatment before breaking the code.

I don't know what this means at all.  What does it mean to "ignores continues"...

> This is the normal situation that we all know today, I did a treatment with an alert ja and when I give an exception within a delegate that execption of a warning in Visual Studio the #pragma would be to disable this warning that Visual Studio and continue the Code because it is an exception that has already been handled

If you want to disable things then uncheck this:

![image](https://user-images.githubusercontent.com/4564579/47243963-a7eb9200-d3a8-11e8-8aba-940dbc897f0e.png)

You may also want to uncheck:

![image](https://user-images.githubusercontent.com/4564579/47243976-b46fea80-d3a8-11e8-84f3-194e88282dd1.png)

--

Your complaints seem to be entirely about the debugger and how it works.  They dont' seem to have any connection to the language at all.

**Comment by franchyd** [comment]
I think he wants a #pragma to disable the debugger breaking on exceptions in certain code paths.

**Comment by CyrusNajmabadi** [comment]
@srburton Debugger requests can be made over at `dotnet/roslyn`.  This doesn't appear to actually involve any sort of language change.
